### PR TITLE
Add DomainMatrix for faster rref in solve

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

This is built on top of #18814 and adds the DomainMatrix class from #18445

#### Brief description of what is fixed or changed

- Add DomainMatrix class
- Use DomainMatrix in solve_linear_system where possible with linsolve as the fallback.

#### Release Notes

* solvers
    * Solving linear systems particularly involving polynomial coefficients is much faster

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
This massively increases the speed of solve when solving linear equations that have symbols in e.g.:
```python
def dosolve(n):
    y = Symbol('y')
    syms = symbols('x:%s' % n)
    eqs = list(y*randMatrix(n, n+1) * Matrix([syms+(1,)]).T)
    sol = solve(eqs, syms)
    return eqs, sol, syms
```
That makes a system like
```julia
In [3]: eqs, sol, syms = dosolve(5)                                                                                                                           

In [4]: eqs                                                                                                                                                   
Out[4]: 
[57⋅x₀⋅y + 14⋅x₁⋅y + 30⋅x₂⋅y + 40⋅x₃⋅y + 50⋅x₄⋅y + 80⋅y, 60⋅x₀⋅y + 50⋅x₁⋅y + 49⋅x₂⋅y + 5⋅x₃⋅y + 5⋅x₄⋅y + 71⋅y, 38⋅x₀⋅y + 59⋅x₁⋅y + 31⋅x₂⋅y + 43⋅x₃⋅y + 58⋅x₄
⋅y + 98⋅y, 92⋅x₀⋅y + 5⋅x₁⋅y + 50⋅x₂⋅y + 64⋅x₃⋅y + 47⋅x₄⋅y + 89⋅y, 83⋅x₀⋅y + 49⋅x₁⋅y + 35⋅x₂⋅y + 50⋅x₃⋅y + 96⋅x₄⋅y + 47⋅y]

In [5]: sol                                                                                                                                                   
Out[5]: 
⎧    120850754      57073183      -239313614       25717805      -72558796 ⎫
⎨x₀: ─────────, x₁: ────────, x₂: ───────────, x₃: ────────, x₄: ──────────⎬
⎩     26138981      26138981        26138981       26138981       26138981 ⎭

In [6]: syms                                                                                                                                                  
Out[6]: (x₀, x₁, x₂, x₃, x₄)
```
On master we have:
```julia
In [7]: for n in range(10): 
   ...:     print() 
   ...:     print(n) 
   ...:     %time ok=dosolve(n) 
   ...:                                                                                                                                                       

0
CPU times: user 384 µs, sys: 406 µs, total: 790 µs
Wall time: 442 µs

1
CPU times: user 9.63 ms, sys: 1.64 ms, total: 11.3 ms
Wall time: 10.1 ms

2
CPU times: user 35.1 ms, sys: 388 µs, total: 35.5 ms
Wall time: 36.3 ms

3
CPU times: user 85.5 ms, sys: 833 µs, total: 86.4 ms
Wall time: 86.8 ms

4
CPU times: user 164 ms, sys: 1.1 ms, total: 165 ms
Wall time: 166 ms

5
CPU times: user 453 ms, sys: 3.71 ms, total: 456 ms
Wall time: 460 ms

6
CPU times: user 2.25 s, sys: 19.3 ms, total: 2.27 s
Wall time: 2.28 s

7
CPU times: user 6.63 s, sys: 34.5 ms, total: 6.66 s
Wall time: 6.72 s

8
CPU times: user 57.9 s, sys: 426 ms, total: 58.4 s
Wall time: 1min 1s

9
^C
```
(I interrupted because it was too slow)

Repeating that with this PR we get
```julia
In [2]: for n in range(10): 
   ...:     print() 
   ...:     print(n) 
   ...:     %time ok=dosolve(n) 
   ...:                                                                                                                                                       

0
CPU times: user 760 µs, sys: 425 µs, total: 1.19 ms
Wall time: 1.11 ms

1
CPU times: user 12.6 ms, sys: 1.87 ms, total: 14.4 ms
Wall time: 13.9 ms

2
CPU times: user 26.4 ms, sys: 917 µs, total: 27.4 ms
Wall time: 27.2 ms

3
CPU times: user 40.5 ms, sys: 440 µs, total: 40.9 ms
Wall time: 41.1 ms

4
CPU times: user 57.8 ms, sys: 422 µs, total: 58.2 ms
Wall time: 58.4 ms

5
CPU times: user 76.5 ms, sys: 652 µs, total: 77.2 ms
Wall time: 77.7 ms

6
CPU times: user 104 ms, sys: 635 µs, total: 105 ms
Wall time: 105 ms

7
CPU times: user 130 ms, sys: 696 µs, total: 130 ms
Wall time: 131 ms

8
CPU times: user 156 ms, sys: 669 µs, total: 157 ms
Wall time: 157 ms

9
CPU times: user 201 ms, sys: 1.04 ms, total: 202 ms
Wall time: 202 ms
```

**Comment by sylee957** [comment]
I think that this approach may start to add a lot of duplicate methods only to operate on a more crude objects.
But it can be worth merging if the speed is improved.

**Comment by oscarbenjamin** [comment]
I don't think that there will need to be many duplicate methods. From discussions elsewhere I think it was concluded that a high-level matrix class could be built from a low-level one that only implemented a few key methods like `rref`, `charpoly` and `det`. My intention here is that we can build something like that and ultimately use it as the basis for all matrix calculations.

The speed improvements are enormous and it also solves the same problems that dotprodsimp doesdoes but in a more natural way. Try the example from https://github.com/sympy/sympy/pull/18814#issuecomment-597151440 to see the speed difference. With the PR we have
```julia
In [12]: eqs, p, y = _mk_eqs(2)                                                                                                   

In [13]: %time ok=solve(eqs, *p.c)                                                                                                
CPU times: user 88.2 ms, sys: 697 µs, total: 88.9 ms
Wall time: 90.2 ms

In [14]: eqs, p, y = _mk_eqs(3)                                                                                                   

In [15]: %time ok=solve(eqs, *p.c)                                                                                                
CPU times: user 400 ms, sys: 1.55 ms, total: 402 ms
Wall time: 404 ms

In [16]: eqs, p, y = _mk_eqs(4)                                                                                                   

In [17]: %time ok=solve(eqs, *p.c)                                                                                                
CPU times: user 1.02 s, sys: 4.96 ms, total: 1.02 s
Wall time: 1.03 s
```
With master we have
```julia
In [2]: eqs, p, y = _mk_eqs(2)                                                                                                    

In [3]: %time ok=solve(eqs, *p.c)                                                                                                 
CPU times: user 962 ms, sys: 2.88 ms, total: 965 ms
Wall time: 965 ms

In [4]: eqs, p, y = _mk_eqs(3)                                                                                                    

In [5]: %time ok=solve(eqs, *p.c)                                                                                                 
CPU times: user 5.42 s, sys: 12 ms, total: 5.44 s
Wall time: 5.45 s

In [6]: eqs, p, y = _mk_eqs(4)                                                                                                    

In [7]: %time ok=solve(eqs, *p.c)                                                                                                 
CPU times: user 34.5 s, sys: 147 ms, total: 34.6 s
Wall time: 35 s
```
Next steps for this PR are to try and use this in place of `PolyMatrix` in the Risch algorithm.

**Comment by sylee957** [comment]
Okay, if it is only a few key methods and not intended to be extended heavily from the mass

**Comment by oscarbenjamin** [comment]
This PR is almost ready but I would like to discuss what exactly to do with the new code that is added. Specifically what should be done with the DomainMatrix class?

The PR adds a DomainMatrix class in sympy/polys/polymatrix.py along with the PolyMatrix class. The DomainMatrix class is based on the polys domains and uses DomainElements as elements. It gives a much faster rref in a number of cases solving many examples that would otherwise hang. However this PR does not put this as a replacement for Matrix.rref. Instead the `solve_lin_sys` function of `sympy/polys/solvers.py` which uses `RawMatrix` is adapted to use `DomainMatrix` instead.

The `solve_lin_sys` function was previously only used by heurisch but is here also used by `solve_linear_system` (and therefore `solve`) and `linsolve` so this unifies solving linear systems of equations apart from the Matrix solving function like `rref` which are themselves no longer used by any of the solvers. A final unification with e.g. `Matrix.rref` seems reasonable but we need to define what the relationship between `Matrix` and `DomainMatrix` should be. Lots of issues involving `Matrix.rref` being slow can be fixed simply by converting to `DomainMatrix` internally but DomainMatrix can also speed up many other things such as matrix multiplication, charpoly, det etc.

Since `solve_lin_sys` no longer uses `RawMatrix` that can be removed which also removes one of the obstacles to ensuring that Matrix can sympify all of its entries. There is also another matrix class that I haven't replaced yet but perhaps should which is the `PolyMatrix` class which is used used in risch (prde.py). On the face of it this has a similar purpose to the DomainMatrix class except is less efficient. The `PolyMatrix` class is a subclass of `Matrix` but having entries that are `Poly` instances i.e. that are `Basic` but not `Expr` (except sometimes they are Poly and sometimes Expr) and like `DomainMatrix` each matrix has an associated domain. It's not really clear to me what the purpose of the `PolyMatrix` class is though. Most likely it could be replaced by `DomainMatrix` except that more of the `Matrix` API might need to be implemented for `DomainMatrix` because of the way that `PolyMatrix` is used.

The other change here is that the `solve_lin_sys` function splits systems according to connected components using the `connected_components` function added in #16299. This gives a significant speed up for large sparse systems of the kind regularly seen in heurisch. This same speedup also carries through to `linsolve` and `solve` as well. Splitting on connected components is also add in `solve` so that it can be used with polynomial systems as well leading to the kinds of speed up discussed here https://github.com/sympy/sympy/pull/16550#issuecomment-529273163

In #18445 there was discussion of adding `DomainMatrix` to be used as an internal object by `Matrix`. The idea discussed there was to make a new immutable `Basic` matrix class that would use a lower-level object like `DomainMatrix` internally. This would make `Matrix` somewhat analogous to `Poly` which is a wrapper around lower-level non-Basic poly instances. It also means that both `Poly` and `Matrix` are based on the same domains so improvements to the domains like adding Gaussian integers can benefit both matrix and polynomial calculations.

The question is if `DomainMatrix` is to be the basis for a new matrix class then it should have various methods added to make it suitable for that purpose. Otherwise it is really just used here for `rref` in which case it does not really need to have many of the methods it currently has like multiplication. On the other hand maybe `DomainMatrix` can be separate from normal `Matrix` and be exposed for end-users who want to do things like matrix calculations over finite fields etc.

Because of the confusion about the relationship between `DomainMatrix` and `Matrix` I am not sure where exactly the code added here belongs. Either way this PR brings enormous speed improvements in many situations so it could be merged as is with the issue of what to do with DomainMatrix deferred until later.

**Comment by sylee957** [comment]
In my experiments, I don't think that sympy objects can wrap around non-sympy objects easily. Because there needs to be an another sympy object that wrap around that object, and the problem just recurses down.

So in my opinion, I think that everything should be done in the basis assuming `DomainMatrix` is some external dependency of `Matrix`.
As mpmath is considered an external dependency and we are using mpmath functions independently when it is needed, the matrix can internally convert to DomainMatrix for appropriate circumstances every time and do computation and convert back to sympy.

I see this approach is also used in `Poly` class, but I think that we can avoid the caveat of having `sympify(Poly(domain=A))` becoming `Poly(domain=B)` if we don't tie `domain` to matrix class itself, but as a keyword argument in function calls.

So we can have the matrix functions with `domain` as optional arguments, like `Matrix.det(domain=ZZ[I])`, `Matrix.rref(domain=FF(2))`.
So the argument can given as a user-provided hint, to skip the decision process of finding the right extension domain where every elements fit in, or doing the computation in other rings.

I also think that this idea is futureproof because even if it can be possible to tie the domain together with Matrix objects, the keyword argument `domain` across each methods can just be made into an redundant argument that forces the original domain matrix have.

The only cons of this is approach is that 
1. Basic arithmetics like multiplication can be done inconveniently if someone wants to use domain, e.g. `A.mul(B, domain=A).mul(C, domain=A)`.
2. I haven't measured the time converting some nested square roots to domain object, but it can be slow.
3. There can be some domain objects unrepresentable in sympy and it can be problem if there is, But I don't think that this is a problem as far as I have observed:
   - Poly <=> Expr
   - FiniteField, ModularInteger <=> Integer
   - AlgebraicElement <=> Sqrt, Pow, ...

**Comment by sylee957** [comment]
The things I find inconvenient in the current implementation of DomainMatrix is that it lacks a lot of convenience methods like multidimensional slicing, submatrix indexing, which often gets handy 
I expect things like `A[1, :] = B[1, :]*C` to be possible also for DomainMatrix and I think that it can ease the translation of algorithms in matrix api to DomainMatrix.
But I think that this can be done carefully as I see indexing done too much complicated in `Matrix`. and I suspect some slowdowns from it.

**Comment by oscarbenjamin** [comment]
At this stage it is deliberate that DomainMatrix is not a full-featured matrix class. I certainly wouldn't want it to have as wide an API as Matrix does but its role needs to be clarified before adding any other API to it. For the purposes here the class itself is not needed and a list of lists/dicts would be perfectly manageable. The question is where we would want to take this in future.

**Comment by sylee957** [comment]
I think that it can be provided as an optional API under Polys after it have lots of self-contained features for operating in a lower level of sympy.
However, I don't want finite field computations to only be possible through DomainMatrix, but not through Matrix. So I think that it should be designed to subset the `Matrix`, and should fill in any disjoint stuff in any way.

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -2553,12 +2555,17 @@ def linsolve(system, *symbols):
...
-            system = [
-                _mexpand(i.lhs - i.rhs if isinstance(i, Eq) else i,
-                recursive=True) for i in system]
-            system, symbols, swap = recast_to_symbols(system, symbols)
-            A, b = linear_eq_to_matrix(system, symbols)
-            syms_needed_msg = 'free symbols in the equations provided'
+            eqs = system
+            try:
+                eqs, ring = sympy_eqs_to_ring(eqs, symbols)
+            except PolynomialError as exc:
```
Comment (smichr) [inline sympy/solvers/solveset.py:2561]: Does this need to recognize ValueError?
```
File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/solvers/solveset.py", line 2525, in sympy.solvers.solveset.linsolve
Failed example:
    linsolve([x**2 - 1], x)
Expected:
    Traceback (most recent call last):
    ...
    NonlinearError:
    nonlinear term encountered: x**2
Got:
    Traceback (most recent call last):
      File "/opt/python/3.8.0/lib/python3.8/doctest.py", line 1328, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.solvers.solveset.linsolve[23]>", line 1, in <module>
        linsolve([x**2 - 1], x)
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/solvers/solveset.py", line 2564, in linsolve
        sol = solve_lin_sys(eqs, ring, _raw=False)
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/polys/solvers.py", line 73, in solve_lin_sys
        eq_coeffs = {ring.gens[monom.index(1)]:coeff for monom, coeff in eq_dict.items()}
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/polys/solvers.py", line 73, in <dictcomp>
        eq_coeffs = {ring.gens[monom.index(1)]:coeff for monom, coeff in eq_dict.items()}
    ValueError: tuple.index(x): x not in tuple
```

Comment (oscarbenjamin) [reply sympy/solvers/solveset.py:2561]: I'm not sure. We have `NonlinearError` now which should probably be raised in this case.

File: sympy/polys/polymatrix.py [hunk 2]
```diff
@@ -87,3 +90,126 @@ def _eval_scalar_rmul(self, other):
 
 
 MutablePolyMatrix = PolyMatrix = MutablePolyDenseMatrix
+
+
+class DomainMatrix:
+
+    def __init__(self, rows, shape, domain):
+        self.shape = shape
+        self.rows = [[item for item in row] for row in rows]
```
Comment (smichr) [inline sympy/polys/polymatrix.py:99]: This is conceptually clashing with Matrix.rows. Could it be private (`_rows`, `_lor` for "list of rows", `dat`, etc..? and accessed through public `tolist()`?)

Comment (smichr) [reply sympy/polys/polymatrix.py:99]: Except for this, I squashed all changes aimed at getting the additions I made and the tests to pass into a single commit past your last commit on my domsolve branch. Please feel free to cherry-pick and modify as you see fit.

**Comment by akirakyle** [comment]
I tried out this PR in hopes of perhaps finding a better faster solver. I'm still trying to grok the many varied solvers sympy has and their pros and cons and perhaps this is a somewhat pathological example but it represents the kind of systems I'm trying to work with but it does seem to illustrate that the current solver in `solve_linear_system` has advantages over linsolve, and other matrix solvers such as LUsolve.
```
a, A, b, B, c, C = symbols("a, A, b, B, c, C", complex=True)
alpha, beta, gamma = symbols("alpha, beta, gamma", complex=True)
Ca, Cb, Cc = symbols("C_a, C_b, C_c", complex=True)
eqns = [A*alpha + a*(gamma + conjugate(gamma)) + Ca,
        B*beta + b*(gamma + conjugate(gamma)) + Cb,
        C*gamma + a*conjugate(alpha) + conjugate(a)*alpha 
        + b*conjugate(beta) + conjugate(b)*beta + Cc]
eqns += [conjugate(eq) for eq in eqns]
alphac, betac, gammac = symbols("alpha^*, beta^*, gamma^*", complex=True)
syms = [alpha, beta, gamma, alphac, betac, gammac]
subs = {conjugate(alpha): alphac,
        conjugate(beta): betac,
        conjugate(gamma): gammac}
eqns = [eq.subs(subs) for eq in eqns]
```
on master `%time soln = solve(eqns, syms)` gives 
```
CPU times: user 30.6 s, sys: 8.38 ms, total: 30.6 s
Wall time: 30.7 s 
```
while this PR gives
```
CPU times: user 2min 5s, sys: 15.8 ms, total: 2min 5s
Wall time: 2min 5s
```
more importantly the solutions are more complex for example `for sym in soln: print(sym, count_ops(soln[sym]))` on master gives
```
alpha 185  
beta 185   
gamma 189  
alpha^* 181
beta^* 181 
gamma^* 184
```
while this PR gives
```
alpha 1347  
beta 1347   
gamma 1171  
alpha^* 1349
beta^* 1349 
gamma^* 1167
```

This PR is certainly an improvement for `linsolve` since I never had the patience for that to even finish on on this example on master. In looking over the code my uneducated guess might be that this new DomainMatrix rref isn't fraction-free by default which the current `solve_linear_system` is, which might explain the large difference in op counts seen.

**Comment by oscarbenjamin** [comment]
Thanks for taking a look. I expect that the difference in op counts is due to the poly domains expanding everything. I think you'll get the same results on master if you call expand at the end.

At the moment DomainMatrix.rref is not division free in this PR. I'm sure I tried that at some point. Right now I can't remember what happened...

You can make it division free something like this:
```julia
diff --git a/sympy/polys/polymatrix.py b/sympy/polys/polymatrix.py
index 28ea1b3602..02401a9282 100644
--- a/sympy/polys/polymatrix.py
+++ b/sympy/polys/polymatrix.py
@@ -203,13 +203,11 @@ def rref(rows, shape):
             continue
         # Normalise row
         pivoti = rows[ri][ci]
-        for ck in range(ci, ncols):
-            rows[ri][ck] = rows[ri][ck] / pivoti
         # Eliminate above and below from col to the right
         for rk in range(nrows):
             pivotk = rows[rk][ci]
             if rk != ri and pivotk:
                 for ck in range(ci, ncols):
-                    rows[rk][ck] = rows[rk][ck] - pivotk * rows[ri][ck]
+                    rows[rk][ck] = pivoti * rows[rk][ck] - pivotk * rows[ri][ck]
         ri += 1
     return rows, pivots
```
I think maybe other changes are needed but in any case this makes it slower rather than faster in your example according to my timings.

Another possibility is that this PR computes the full rref but it would be better to compute the echelon form (possibly division free) and then do back substitution with normal sympy types to avoid some of the expansion.

**Comment by akirakyle** [comment]
Indeed you are correct that the op count difference is simply due to the `DomainMatrix` solution being in expanded form. I probably should have checked that before posting my comment. However I believe the performance difference is still large enough to warrant reworking this to be on par with the current `solve_linear_system`. Testing the fraction-free patch you proposed, I also see a significant slow down and the resulting expressions are blown up with tens of thousands of ops. Looking at the current implementation of `solve_linear_system` it seems that it does only compute echelon form then do back substitution so I think it is worth considering trying that. I would take a stab at it, but I don't have any knowledge of the polynomial portion of this.

**Comment by akirakyle** [comment]
For reference, it is the elimination loop that takes the majority of the time here, and that time scales fairly poorly
```
ci: 0   elim time: 0.17379283905029297
ci: 1   elim time: 0.3668661117553711
ci: 2   elim time: 5.571394443511963
ci: 3   elim time: 14.886268377304077
ci: 4   elim time: 31.004244327545166
ci: 5   elim time: 73.20919919013977
```
which is timing this block
```
        for rk in range(nrows):
            pivotk = rows[rk][ci]
            if rk != ri and pivotk:
                for ck in range(ci, ncols):
                    rows[rk][ck] = rows[rk][ck] - pivotk * rows[ri][ck]
```

**Comment by oscarbenjamin** [comment]
It is at least expected that the majority of the time would be spent there as that's the inner loop of the algorithm. I assume that the `ci:0` etc are referring to the time for each iteration of the loop in the example that you showed. The fact that each iteration takes longer is because the expressions are growing more and more complex.

I think ultimately this PR is just slower because in your example the expressions are large when expanded. That is the ultimate problem that I haven't really worked out from this PR. The poly domains are efficient compared to Basic when you have low order polynomials in the coefficients of a linear system. They can be absurdly inefficient if some high-order expression is unnecessarily expanded. Actually solve already suffers from that problem on master because it uses Poly to get the coefficients before calling solve_linear_system. Here's a simple example
```julia
In [15]: solve([y - (x+1)**10, z], [y, z])                                                                                                     
Out[15]: 
⎧    10       9       8        7        6        5        4        3       2                 ⎫
⎨y: x   + 10⋅x  + 45⋅x  + 120⋅x  + 210⋅x  + 252⋅x  + 210⋅x  + 120⋅x  + 45⋅x  + 10⋅x + 1, z: 0⎬
⎩                                                                                            ⎭
```
If you increase the 10 then this just gets slower and the output gets bigger.

**Comment by akirakyle** [comment]
After playing around with this some more I think you are correct that the majority of the slow behavior for this PR is due to the poly domains. However I think the inherent algorithmic difference with the current `solve_linear_system` might also matter, especially if the polys domains is made faster on this sort of "higher-order" polynomials in the coefficients (although the solutions here only has terms of up to order 2 in any single coefficient but order 6 as a set and I guess there are a lot of terms). 

Specifically I tried implementing what `solve_linear_system` does (and what you also suggested) which is to compute the echelon form and then do back substitution with normal sympy types. Here's my modified rref:
```python
def rref(rows, shape, normalize_last=False, normalize=False, zero_above=False):
    import time
    nrows, ncols = shape
    rows = [[item for item in row] for row in rows]
    pivots = []
    ri = 0
    for ci in range(ncols):
        for rj in range(ri, nrows):
            if rows[rj][ci]:
                # Row swap for pivot
                if rj != ri:
                    rows[rj], rows[ri] = rows[ri], rows[rj]
                # Record pivot
                pivots.append(ci)
                break
        else:
            # No pivot
            continue
        # Normalise row
        pivoti = rows[ri][ci]
        if normalize and not normalize_last:
            t0 = time.time()
            for ck in range(ci, ncols):
                rows[ri][ck] = rows[ri][ck] / pivoti
            pivoti = rows[ri][ci]
            t1 = time.time()
            print("ci:", ci, "  norm time:", t1 - t0)
        t2 = time.time()
        # Eliminate above (if zero_above) and below from col to the right
        for rk in range(0 if zero_above else ri+1, nrows):
            pivotk = rows[rk][ci]
            if rk != ri and pivotk:
                for ck in range(ci, ncols):
                    #rows[rk][ck] -= pivotk * rows[ri][ck]
                    rows[rk][ck] = pivoti * rows[rk][ck] - pivotk * rows[ri][ck]
        ri += 1
        t3 = time.time()
        print("ci:", ci, "  elim time:", t3 - t2)
    if normalize_last:
        for ci in reversed(pivots):
            t0 = time.time()
            ri -= 1
            pivoti = rows[ri][ci]
            for ck in range(ci, ncols):
                rows[ri][ck] = rows[ri][ck] / pivoti
            t1 = time.time()
            print("ci:", ci, "  norm last time:", t1 - t0)
        assert(ri == 0)
    return rows, pivots
```
and here's my modified  `_solve_lin_sys_component`:
```python
def _solve_lin_sys_component(eqs_coeffs, eqs_rhs, ring):
    import time

    # transform from equations to matrix form
    matrix = eqs_to_matrix2(eqs_coeffs, eqs_rhs, ring.gens, ring.domain)

    # solve by row-reduction
    echelon, pivots = matrix.rref()

    # construct the returnable form of the solutions
    keys = ring.gens

    if pivots and pivots[-1] == len(keys):
        return None

    sols = {}
    echelon = echelon.rows
    for i, p in enumerate(pivots):
        t0 = time.time()
        v = sum(echelon[i][j]*keys[j] for j in range(p+1, len(keys)))
        sols[keys[p]] = (echelon[i][-1] - v)/echelon[i][p]
        print("sols:", i, "  p:", p, "  time:", time.time() - t0)

    return sols
```
which `%time soln = solve(eqns, syms)` gives me
```
ci: 0   elim time: 0.1441974639892578
ci: 1   elim time: 0.18923497200012207
ci: 2   elim time: 0.8090705871582031
ci: 3   elim time: 1.4696333408355713
ci: 4   elim time: 7.784857988357544
ci: 5   elim time: 4.5299530029296875e-06
[0, 1, 2, 3, 4, 5]
sols: 0   p: 0   time: 0.04331088066101074
sols: 1   p: 1   time: 0.02683424949645996
sols: 2   p: 2   time: 0.40357208251953125
sols: 3   p: 3   time: 0.615847110748291
sols: 4   p: 4   time: 1.3736724853515625
sols: 5   p: 5   time: 22.65699052810669
CPU times: user 35.7 s, sys: 73.6 ms, total: 35.8 s
Wall time: 35.9 s
```
However I'm really cheating with these timings since I haven't actually performed the back substitution yet since I just wanted `solve_lin_sys` to convert the echelon form equations back to `Expr`. Now I can do something like:
```python
import time
t0 = time.time()
soln_subs = {}
for k,v in reversed(list(soln.items())):
    soln_subs[k] = v.subs(soln_subs)
print("time:", time.time() - t0)
```
Which only takes ~3 seconds, however the resulting expressions are very large and simplifying them with `ratsimp` and `.expand()` takes longer than solving the equations. Which now I've come full circle which is that somehow `solve_linear_system` manages to be both faster and return simpler results than the `DomainMatrix` solution for this example. In fact turning simplification off in `solve_linear_system` gives me the solution in ~1 sec on my machine still with fewer ops than the `DomainMatrix` solver.

I hope I'm not coming across as being down on this `DomainMatrix` solver. I really like it and as I've been trying to understand the solvers in sympy and their advantages and disadvantages I've seen how much of a mess it is with similar functionality implemented several times across several classes.  I hope having this example helps!

**Comment by oscarbenjamin** [comment]
Thanks @akirakyle for looking at this. That is helpful. I would have responded earlier but I just haven't had time to look at this.

I think what I need to do with this is split it up to get the more straight-forward parts merged and then focus on the trickier bits after.

**Comment by sylee957** [comment]
I notice that Poly with single entries can be a representative for domain elements, so I'd rather view this as a trivial form of `PolyMatrix` which only stores Poly with 0 degrees.
Do you have some rigorous reason to replace this in place of `PolyMatrix`?

**Comment by oscarbenjamin** [comment]
> Do you have some rigorous reason to replace this in place of `PolyMatrix`?

A Poly always has to have generators. More importantly PolyMatrix is based on simply putting a non-Expr Poly in a Matrix (and consequently doesn't really work properly). It's an example of the kind of fudged implementation that appears a lot around sympy which on the one hand limits improvements to the matrices code and on the other hand gives a PolyMatrix class that barely works.

DomainMatrix is a completely separate class based directly on the domains and sharing no code with Matrix so it doesn't have this baggage. Also ground domains like QQ can be a lot faster than a Poly over QQ and can support division even if Poly can not. Potentially we could make DomainMatrix a Basic subclass so it would have a similar role/interface as Poly.

When there are symbols in the expressions the representation here is based on the sparse polys rather than the dense representation used by Poly (which is inefficient in many cases).

I would rather think that this class could be built up to something that works as a replacement for PolyMatrix or can be something that PolyMatrix could be implemented over the top of. Ultimately I think that Matrix should use something like this internally for all calculations like rref etc.

**Comment by sylee957** [comment]
I think that this inspires to try defining the `DomainMatrix.args` to be entries of Poly instances with constants with equal domains and internally it can have an another internal array representation used for computation. 
So `DomainMatrix([[Poly, ...], [Poly, ...], ...])` is how it's going to be created outside and `DomainMatrix._mat = [DomainElement, DomainElement, ...]` is how it's going to be internally represented.
Have you considered this idea?

**Comment by oscarbenjamin** [comment]
Remember that there is no way to represent a Poly without generators so if the matrix is just all rational numbers then what would the args be?

**Comment by sylee957** [comment]
It can be represented by having an arbitrary bound symbol.

**Comment by oscarbenjamin** [comment]
I don't see any advantage in using Poly here. If we returned a Poly object from e.g. `M[i, j]` then it would always be a "constant" polynomial and none of the Poly methods would do anything useful.

I think it's better to see DomainMatrix as being a low-level class like the domain elements. It's more analogous to the non-Basic DMP objects used internally by Poly than it is to Poly itself. If we want to make it more convenient for users then we can wrap it in a high-level class. It would be better to keep the low-level class very simple though: its API should be designed to optimise performance and ease of implementation rather than ease of use.

**Comment by oscarbenjamin** [comment]
The benchmark results for this are looking good:
https://travis-ci.org/github/sympy/sympy/jobs/701861860#L1406

I want to clean up the commits before merging but otherwise this is ready for review if anyone wants to take a look.

I really think it would be good to get this in and think about how to build up `DomainMatrix` to be something that can make general matrix computation faster especially after the dotprodsimp revert (#19588). This has the potential to massively speed up many matrix calculations e.g. https://github.com/sympy/sympy/issues/19532#issuecomment-643219232

**Comment by jksuom** [comment]
Would it be possible to make `DomainMatrix` a subclass of `MatrixBase` (or some existing matrix class)? Then it might not be necessary to redefine all methods like colon and row operations, `nullspace`, `eye`, `zeros`, ...

**Comment by oscarbenjamin** [comment]
I would rather keep the implementation of `DomainMatrix` completely separate from the existing matrix classes. I expect that the tangled hierarchy of classes used to make Matrix would only lead to complications or make some things inefficient. We could make a higher-level class that wraps `DomainMatrix` and provides an interface similar to other matrix classes though.

**Comment by sylee957** [comment]
I also think that it should be PolyMatrix which should really resolve the gap between DomainMatrix and Matrix.

**Comment by oscarbenjamin** [comment]
This PR is now ready for review or merge

**Comment by oscarbenjamin** [comment]
Thanks @sylee957 and thanks to everyone else for reviewing and helping

**Comment by moorepants** [comment]
Hi, this PR causes output in matrix operations to result in over complicated results and has an effect on the physics mechanics codes.

See this issue for an explanation: https://github.com/sympy/sympy/issues/23130

This commit causes the regression: f0b7996ca517e3b78749ec5c40e86d9d6efc0f23

The results are mathematically equivalent but the expressions are longer and introduce significantly longer equations in complex multi-body systems (i.e. after passed through KanesMethod/LagrangesMethod). This is an undesired effect. It is a case were the output, even though mathematically the same, shouldn't have changed.

**Comment by moorepants** [comment]
Is there an option to not use this DomainMatrix underlying implementation? If so, I can disable it for the physics vector/mechanics codes.

**Pull request merged**